open_spiel/games/universal_poker.cc

```diff
@@ -466,33 +524,216 @@ std::unique_ptr<State> UniversalPokerState::Clone() const {
   return absl::make_unique<UniversalPokerState>(*this);
 }
 
+
+int GetHoleCardsReachIndex(int card_a, int card_b,
+                           int num_suits, int num_ranks) {
+  // CardSet uses kSuitChars of "cdhs", but we need the inverse i.e. "shdc"
+  // according to https://github.com/Sandholm-Lab/LibratusEndgames
+  const int a_suit = num_suits - 1 - suitOfCard(card_a);
+  const int b_suit = num_suits - 1 - suitOfCard(card_b);
+  const int a_rank = rankOfCard(card_a);
+  const int b_rank = rankOfCard(card_b);
+  // Order by rank, then order by suit.
+  const int lesser_card = a_rank < b_rank
+                              || (a_rank == b_rank && a_suit < b_suit)
+                          ? card_a : card_b;
+  // We use ints, so this is ok.
+  const int higher_card = card_a + card_b - lesser_card;
+
+  const int lesser_suit = num_suits - 1 - suitOfCard(lesser_card);
+  const int higher_suit = num_suits - 1 - suitOfCard(higher_card);
+  const int lesser_rank = rankOfCard(lesser_card);
+  const int higher_rank = rankOfCard(higher_card);
+
+  // There is an ordering on the cards, which indexes an upper triangular matrix
+  // like so (example for num_cards_in_deck=4)
+  //
+  //       j ->
+  //     0 1 2 3
+  //   0 # 0 1 2
+  // i 1   # 3 4
+  // | 2     # 5
+  // v 3       #
+  //
+  const int i = lesser_rank * num_suits + lesser_suit;
+  const int j = higher_rank * num_suits + higher_suit;
+  const int n = num_suits * num_ranks;
+  // Compute index k in the triangle.
+  return i * (2*n - i - 3) / 2 + j - 1;
+}
+
+// Normally distribution of cards happens in a tree, one card at a time to each
+// player. We flatten this distribution and omit repetitions.
+// The yield function is called for each unique tuples of cards the players can
+// receive.
+void DistributeHands(const logic::CardSet& full_deck,
+                     uint8_t num_players,
+                     uint8_t num_hole_cards,
+                     std::function<void(/*per_player_hole_cards=*/
+                            std::vector<std::vector<uint8_t>>)> yield) {
+  // Taken and modified from CardSet::SampleCards()
+  int nbCards = num_hole_cards * num_players;
+
+  uint64_t p = 0;
+  for (int i = 0; i < nbCards; ++i) {
+    p += (1 << i);
+  }
+  // Enumerates all the uint64_t integers that with nbCards 1-bits.
+  // The final n is ignored. It is fine as long as the rank < 16.
```

Comment:
there should be a check here to validate rank<16, if not appropriate error message should be thrown

Reply:
Cards with rank >= 16 are not representable in CardSet, so there is nothing to check. I updated the comment.